pkg/lightning/backend/local/duplicate.go

```diff
@@ -0,0 +1,636 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package local
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"sort"
+	"sync"
+	"time"
+
+	split "github.com/pingcap/br/pkg/restore"
+
+	"github.com/pingcap/kvproto/pkg/kvrpcpb"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/distsql"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/ranger"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/keepalive"
+
+	backendkv "github.com/pingcap/br/pkg/lightning/backend/kv"
+	"github.com/pingcap/br/pkg/lightning/common"
+	"github.com/pingcap/br/pkg/lightning/log"
+	"github.com/pingcap/br/pkg/logutil"
+
+	"github.com/cockroachdb/pebble"
+	"github.com/pingcap/errors"
+	sst "github.com/pingcap/kvproto/pkg/import_sstpb"
+	kvrpc "github.com/pingcap/kvproto/pkg/kvrpcpb"
+	tikv "github.com/pingcap/kvproto/pkg/tikvpb"
+)
+
+const (
+	maxWriteBatchCount = 128
+)
+
+type DuplicateRequest struct {
+	tableID   int64
+	indexID   int64 // 0 represent it is a table request
+	start     kv.Key
+	end       kv.Key
+	indexInfo *model.IndexInfo
+}
+
+type DuplicateManager struct {
+	db                *pebble.DB
+	splitCli          split.SplitClient
+	regionConcurrency int
+	connPool          common.GRPCConns
+	tls               *common.TLS
+	sqlMode           mysql.SQLMode
+	ts                uint64
+	keyAdapter        KeyAdapter
+}
+
+func NewDuplicateManager(
+	db *pebble.DB,
+	splitCli split.SplitClient,
+	ts uint64,
+	tls *common.TLS,
+	regionConcurrency int,
+	sqlMode mysql.SQLMode) (*DuplicateManager, error) {
+	return &DuplicateManager{
+		db:                db,
+		tls:               tls,
+		regionConcurrency: regionConcurrency,
+		sqlMode:           sqlMode,
+		splitCli:          splitCli,
+		keyAdapter:        duplicateKeyAdapter{},
+		ts:                ts,
+		connPool:          common.NewGRPCConnc(),
+	}, nil
+}
+
+func (manager *DuplicateManager) DuplicateTable(ctx context.Context, tbl table.Table) error {
+	log.L().Info("Begin collect duplicate data from remote TiKV")
+	reqs, err := buildDuplicateRequests(tbl.Meta())
+	if err != nil {
+		return err
+	}
+
+	decoder, err := backendkv.NewTableKVDecoder(tbl, &backendkv.SessionOptions{
+		SQLMode: manager.sqlMode,
+	})
+	if err != nil {
+		return err
+	}
+	var wg sync.WaitGroup
+	var tableErr common.OnceError
+	rpcctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	for _, r := range reqs {
+		wg.Add(1)
+		go func(req *DuplicateRequest) {
+			err := manager.sendRequestToTiKV(rpcctx, decoder, req)
+			if err != nil {
+				log.L().Error("error occur when collect duplicate data from TiKV", zap.Error(err))
+				tableErr.Set(err)
+				cancel()
+			}
+			wg.Done()
+		}(r)
+	}
+	wg.Wait()
```

Comment:
How about `errgroup.Group` ?

The code looks like:
```go
	g, rpcctx := errgroup.WithContext(ctx)
	for _, r := range reqs {
		req := r
		g.Go(func() error {
			err := manager.sendRequestToTiKV(rpcctx, decoder, req)
			if err != nil {
				log.L().Error("error occur when collect duplicate data from TiKV", zap.Error(err))
			}
			return err
		})
	}
	err := g.Wait()

```

Reply:
done.